ambry-store/src/main/java/com/github/ambry/store/BootstrapController.java

```diff
@@ -0,0 +1,284 @@
+import java.util.regex.Pattern;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.clustermap.StateTransitionException.TransitionErrorCode.*;
+
+
+/**
+ * BootstrapController is responsible for managing the hydration protocol
+ * for Offline -> Bootstrap state transition of partitions.
+ */
+public class BootstrapController {
+  private boolean isRunning = false;
+  private final String BOOTSTRAP_IN_PROGRESS_FILE_NAME;
+  private final String FILECOPY_IN_PROGRESS_FILE_NAME;
+  private final Pattern allLogSegmentFilesPattern = Pattern.compile(".*_log.*");
+  private final ServerConfig serverConfig;
+  private final StoreManager storeManager;
+  private final PartitionStateChangeListener storageManagerListener;
+  private final PartitionStateChangeListener fileCopyManagerListener;
+  private final ClusterParticipant primaryClusterParticipant;
+
+  private static final Logger logger = LoggerFactory.getLogger(BootstrapController.class);
+
+  public BootstrapController(
+      @Nonnull StoreManager storeManager, @Nonnull StoreConfig storeConfig,
+      @Nonnull ServerConfig serverConfig, @Nonnull ClusterParticipant primaryClusterParticipant) {
+    this.serverConfig = serverConfig;
+    this.storeManager = storeManager;
+    this.primaryClusterParticipant = primaryClusterParticipant;
+
+    this.BOOTSTRAP_IN_PROGRESS_FILE_NAME = storeConfig.storeBootstrapInProgressFile;
+    this.FILECOPY_IN_PROGRESS_FILE_NAME = storeConfig.storeFileCopyInProgressFileName;
+
+    primaryClusterParticipant.registerPartitionStateChangeListener(
+        StateModelListenerType.BootstrapControllerListener, new BootstrapControllerImpl());
+    logger.info("Bootstrap Controller's state change listener registered!");
+
+    Map<StateModelListenerType, PartitionStateChangeListener> partitionStateChangeListeners =
+        primaryClusterParticipant.getPartitionStateChangeListeners();
+
+    this.storageManagerListener =
+        partitionStateChangeListeners.get(StateModelListenerType.StorageManagerListener);
+    this.fileCopyManagerListener =
+        partitionStateChangeListeners.get(StateModelListenerType.FileCopyManagerListener);
+  }
+
+  void start() {
+    if(!isRunning) {
+      //Start the FileStore
+      isRunning = true;
+    }
+  }
+
+  void shutdown() {
+    //Implement shutdown Hook.
+    isRunning = false;
+  }
+
+  boolean isRunning() {
+    return isRunning;
+  }
+
+  class BootstrapControllerImpl implements PartitionStateChangeListener {
+    EnumSet<ReplicationProtocolTransitionType> replicationProtocolTransitionType;
+
+    /**
+     * TODO: Add context
+     * @param partitionName of the partition.
+     */
+    @Override
+    public void onPartitionBecomeBootstrapFromOffline(@Nonnull String partitionName) {
+      logger.info("Bootstrap Controller's state change listener invoked for partition `{}`, state change `{}`",
+          partitionName, "Offline -> Bootstrap");
```

Comment:
nit: "offline->bootstrap" is not a variable, can just be put in the log message directly.

Reply:
Incorporated.